### DISCUSSION TITLE
How to put MP onto ItsyBitsy nRF52840?

### DISCUSSION THREAD

**Discussion by rkompass** [micropython/micropython#13482] [nRF]
I'm asking for help:
Can someone share his procedure to put Micropython onto the (ItsyBitsy) nRF52840?
I'm assuming that the firmware of the [Feather nRF52840 Express](https://micropython.org/download/FEATHER52/) should work as it is the same processor (and bootloader).
However at our download page there are only .bin and .hex files.
I managed to convert the .hex file with
`uf2conv.py -c -f 0xADA52840 -b 0x27000 -o feather.uf2 FEATHER52-20240117-v1.23.0-preview.47.g16c6bc47c.hex`
which reports:
```
Converted to uf2, output size: 371712, start address: 0x0
Wrote 371712 bytes to feather.uf2
```
The resulting .uf2 put onto the NRF52BOOT drive should install, but after RESET I cannot communicate to the Board, there is no ttyACM0 or similar.

With another board ( SEEED XIAO nRF52840 Sense ) there are installation instructions and they tell that after using 
`uf2conv.py -c -f 0xADA52840 -o firmware.uf2 firmware.hex` it must report the start address as 0x27000.
Which is not the case here.
Could that be the culprit? How can I change the start address then?
Or are there other reasons?

**Comment by cwalther** [comment]
The board definitinon named [FEATHER52](https://github.com/micropython/micropython/blob/master/ports/nrf/boards/FEATHER52/mpconfigboard.mk) and labeled as “Feather nRF52840 Express” in board.json is actually for an nRF52832, so I’m not sure if it would run on an ItsyBitsy nRF52840 (or even on the Feather).

What I did is make [my own board definition for the ItsyBitsy nRF52840](https://github.com/cwalther/dsd6-micropython/tree/main/boards/ITSYBITSY_NRF52840). I have a couple of [pending pull requests for nRF](https://github.com/micropython/micropython/pulls?q=is%3Apr+author%3Acwalther+nrf), at least #13071 is needed to use that board definition. I have built working UF2 firmware from that. I suppose I could build you one of the specific version you want, if you are not set up to build it yourself.

**Comment by robert-hh** [comment]
@rkompass  The location of the firmware depends on the softdevice version which is being used. And the firmware has to be built accordingly. The load address can be 0x26000 (SD 6.0.1), 0x27000 (SD 7.x).

**Comment by rkompass** [comment]
Thank you both. I had a look at the build instructions: There is the Bluetooth Stack (SD) and that is not the soft device, is it?

> The location of the firmware depends on the softdevice version which is being used. And the firmware has to be built accordingly.

I assume that the `bootloader.ld` address has to match the definitions in `mpconfigport.mk`? And trust in that you configured it right @cwalther . I will try to build it later.

Do you have experience with the [BOSSA flash programming utility ](https://learn.adafruit.com/adafruit-metro-m0-express/uf2-bootloader-details) ? That would allow flashing the bin file without conversion to `.uf2`.

**Reply by robert-hh** [reply]
The Soft device code is in .../nrf/drivers/bluetooth for various versions, but it is AFAIK not loaded with the firmware. The Adafruit board comes with a soft device, so the code has to match it. The Adafruit memory map is defined in https://github.com/adafruit/circuitpython/blob/main/ports/nrf/mpconfigport.h, line 102 and up. The code has to match the SD. You may get some additional hints by looking at the board files for Seeed XIAO NRF52840.

bossac is used e.g. for the `Arduino nano 33 BLE sense`. But it may need a specific bootloader at the device.

When adding the Seeed XIAO NRF52840 board, I wiped the bootloader. In the load instructions for this there is a link on how to reflash the bootloader. 

P.S.: It's a few months since I worked on the nrf bootloader part, and meanwhile I forget things fast. My notes are also not very detailed.

**Reply by cwalther** [reply]
> There is the Bluetooth Stack (SD) and that is not the soft device, is it?

As far as I understand (but I’m a newbie at this too): The softdevice (SD) is some kind of hardware abstraction layer or rudimentary operating system that includes the Bluetooth stack. It is installed in a separate flash range from the application and typically not included in the application binary (but sometimes included in the bootloader binary). For SD 140 (that number says what hardware it is for, nRF52840) version 6.x, that is 0x0 to 0x26000.

> I assume that the `bootloader.ld` address has to match the definitions in `mpconfigport.mk`?

No, _bootloader.ld_ is there to carve out space at the end of the flash for the bootloader (consisting of sub-parts bootloader MBR, bootloader code, bootloader settings), otherwise that would be overwritten by the filesystem. That’s a mistake I made in the first version of my board definition.

The values set in _bootloader.ld_ and _nrf/boards/nrf52840_1M_256k.ld_ are inserted into the definitions from _nrf/boards
/memory.ld_ and the softdevice and bootloader boundaries in the memory map formed by that must match what’s in the CircuitPython file linked by robert-hh.

> And trust in that you configured it right @cwalther.

Well, it’s empirically right enough to run, but there may still be mistakes (like the one about overwriting the bootloader with the filesystem, mentioned above, that is now fixed). So corrections are welcome.

> Do you have experience with the [BOSSA flash programming utility](https://learn.adafruit.com/adafruit-metro-m0-express/uf2-bootloader-details)?

No, I don’t have that installed (except probably buried somewhere in the Arduino IDE) and I find UF2 the most convenient way of loading firmware, so I didn’t try anything else.

**Reply by andrewleech** [reply]
Yeah the SoftDevice (SD) is the way NRF always used to add bluetooth support, it's a closed source binary blob that needs to be flashed to the right location, and then your application needs to be configured / linked just right for that version of SD to share the flash / ram properly.

It's a huge pain to deal with, and knowing how much ram to share with the SD is trial and error depending on how many BLE services & characteristics you might want to support.

Also, bluetooth using SD is _not_ compatible with the "modern" micropython BLE api, nor with aioble.

I started work porting the nimble stack (as used on other micropython ble devices) over to nrf port here: https://github.com/micropython/micropython/pull/8318 but the project I was using it on got paused... then when I started it again I switched to rpi pico w instead so never got back to this.

I have heard there's a long term desire to support nrf better with the zephyr port because nordic themselves are contributing to zephyr so most of their newer ble features only really work there. I think there's a lot of work to go to get that up and running properly though.

**Reply by cwalther** [reply]
Speaking of Bluetooth APIs – do you know whether the the `ubluepy` API supports HID (acting as a mouse or keyboard)? I know next to nothing about BLE, but after reading the first bits here and there I have a hunch it might not. Which would be a bummer, because that was the whole reason I got into getting MicroPython on my nRF52832 fitness tracker.

**Reply by andrewleech** [reply]
@cwalther I'm not sure if hid requires different descriptors or something that might not be easily supported? You would probably find it easier to use my nimble support branch and then use aioble?

**Reply by cwalther** [reply]
I need to read up on this more thoroughly, but https://www.bluetooth.com/specifications/specs/human-interface-device-profile-1-1-1/ seems to suggest that HID requires L2CAP, of which I see no mention in ubluepy (in contrast to the [bluetooth](https://docs.micropython.org/en/latest/library/bluetooth.html#l2cap-connection-oriented-channels) module). But now I’m confused whether that specification is about BLE at all, or Bluetooth Classic.

At any rate, if your branch is complete enough for that, I’ll give it a try!

**Reply by rkompass** [reply]
@andrewleech This sounds like your branch allows for using aioble already?
That's exciting. I saw that Circuitpython pulled your branch in.
All I want is to have a similar experience like on the other MP boards that support Bluetooth.
As I'm learning all this freshly I'm not interested in a lot of different APIs, of course.

How much work is to be done in your branch yet? Can someone not acquainted with Bluetooth yet help?

**Reply by andrewleech** [reply]
I rebased my branch last night, it was a fair bit harder than I expected there's been a reasonable amount of change to the nrf port in the last couple of years.

I've done just a basic test of the aioble temp_sensor example and that worked on my 52840 dongle, not sure if I've got any of the older / smaller chips around still to test on.

I'll need to re-run the multitests at some stage...

**Reply by rkompass** [reply]
I have to confess I'm a Github layman still, unfortunately.
Can someone (of you?) combine the timer/RTC PRs and your last branch (of @andrewleech).
So I can clone this and compile?
I might try to run the multitests perhaps (if they are not too complicated).

**Comment by rkompass** [comment]
Thanks to your help I have a little more insight now.
I could compile the firmware copying your (@cwalther) board definitions into the `ports/nrf/boards folder`.
Then I converted the `firmware.hex` with `uf2conv.py firmware.hex -c -f 0xADA52840`.
The start address reported is 0x26000 which seems good.

Now I'm stuck with an old or perhaps **wrong?** bootloader apparently.
The `INFO_UF2.TXT` contains:
```
UF2 Bootloader 0.5.0 lib/nrfx (v2.0.0) lib/tinyusb (0.9.0-22-g7cdeed54) lib/uf2 (remotes/origin/configupdate-9-gadbb8c7)
Model: Nordic nRF52840 Dongle
Board-ID: nRF52840-pca10059-v1
SoftDevice: S140 version 6.1.1
Date: Apr  4 2021
```
which is a little progress after I put a converted to .uf2 [https://github.com/adafruit/Adafruit_nRF52_Bootloader/releases/download/0.8.2/itsybitsy_nrf52840_express_bootloader-0.8.2_s140_6.1.1.hex](https://github.com/adafruit/Adafruit_nRF52_Bootloader/releases/download/0.8.2/itsybitsy_nrf52840_express_bootloader-0.8.2_s140_6.1.1.hex) onto the boot drive.
Before that it told SoftDevice: none

Still the bootloader is version 0.5 and this version does not allow to flash big firmwares as needed.
Trying to update the bootloader with `update-itsybitsy_nrf52840_express_bootloader-0.8.2_nosd.uf2` from the adafruit [github](https://github.com/adafruit/Adafruit_nRF52_Bootloader/releases)
does not change it, apparently.

Could it be that I have a completely wrong bootloader (thinking to be a Nordic nRF52840 Dongle, see the above file content), that refuses to flash a firmware for another board?

Also I'm confused: Apparently the bootloader was not updated by the `itsybitsy_nrf52840_express_bootloader-0.8.2_s140_6.1.1.uf2` yet the SoftDevice was?

**Reply by rkompass** [reply]
Update: I see now the compilation was o.k.. because I could flash my firmware onto a second board:
```
MicroPython v1.23.0-preview.56.gf7f53f2b9.dirty on 2024-01-23; Adafruit ItsyBitsy nRF52840 Express with NRF52840
Type "help()" for more information.
```
but on that board was this bootloader:
```
UF2 Bootloader 0.7.0 lib/nrfx (v2.0.0) lib/tinyusb (0.12.0-145-g9775e7691) lib/uf2 (remotes/origin/configupdate-9-gadbb8c7)
Model: Adafruit ItsyBitsy nRF52840 Express
Board-ID: nRF52840-ItsyBitsy-revA
Date: Jul 13 2022
SoftDevice: S140 6.1.1
```
So this is at least partial success. I shall incorporate you pull requests @cwalther :-)

And now only this question remains: How do I get the right bootloader on my first board?
Thank you all for your help, btw.

Raul

**Reply by robert-hh** [reply]
When I wrecked the bootloader one on a Seeed device, I could restore it with my jlink adapter. You "just" need the hex file with the S140 6.1.1 soft device. Maybe you can copy it from the working board. Just get the address range up to 0x26000.

**Reply by rkompass** [reply]
Thank you robert.
I just now found:
`adafruit-nrfutil --verbose dfu serial --package itsybitsy_nrf52840_express_bootloader-0.8.2_s140_6.1.1.zip -p /dev/ttyACM0 -b 115200 --singlebank --touch 1200` and the bootloader got updated.
I could flash my firmware afterwards.

**Reply by cwalther** [reply]
Good to hear that you got it sorted out. FWIW, on my board I have the same 0.7.0 as on your second one, and I had installed that using _update-itsybitsy_nrf52840_express_bootloader-0.7.0_nosd.uf2_. (I’m not sure why I hadn’t downloaded 0.8.x, maybe I just followed the link from the CircuitPython download page.)

**Reply by rkompass** [reply]
It seems the serial update with adafruit-nrfutil is more universal in that you may flash a bootloader that is even for another board, so I needed that.

Just noticed there is no `machine.freq()`.
